### PR TITLE
Simplify halo plugin manager

### PR DESCRIPTION
**Pull request by JohnNiang**
#### What type of PR is this?

/kind improvement
/area core
/area plugin
/milestone 2.12.x

#### What this PR does / why we need it:

This PR mainly simplifies halo plugin manager. Before this,
- we have too many repeat code from super class, which is uncessary
- we maintain plugin application context in ExtensionComponentsFinder, which is uncessary and is hard to manage
- we fire halo plugin event in halo plugin manager, which is complicated and leads to too many repeat code

This PR does:
- refactor halo plugin manager
- wrap base plugin with spring plugin which contains application context
- remove ExtensionComponentsFinder
- bridge halo plugin event and spring plugin event
- wait extensions fully deleted when stopping

Meanwhile, this PR will supersede PR <https://github.com/halo-dev/halo/pull/5236>.

#### Which issue(s) this PR fixes:

Fixes https://github.com/halo-dev/halo/issues/5226

#### Special notes for your reviewer:

Test installing, enabing, disabling, upgrading, reloading and deleting plugins.

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: application/src/main/java/run/halo/app/plugin/PluginApplicationContext.java [hunk 1]
```diff
@@ -16,18 +16,18 @@
  * @author guqing
  * @since 2.0.0
  */
-public class PluginApplicationContext extends GenericApplicationContext {
+public class PluginApplicationContext extends AnnotationConfigApplicationContext {
```
Comment (guqing) [inline application/src/main/java/run/halo/app/plugin/PluginApplicationContext.java:19]: 这里继承了 AnnotationConfigApplicationContext 我认为是没有必要的，其中的 scan 过程会很耗时，而且是不需要的，插件构建时就已经完成了对 bean 扫描和索引会在创建 PluginApplicationContext 时直接传递给它

Comment (JohnNiang) [reply application/src/main/java/run/halo/app/plugin/PluginApplicationContext.java:19]: 我刚刚 debug 了一下，实际上并不会调用 `org.springframework.context.annotation.AnnotationConfigApplicationContext#scan`。

Comment (guqing) [reply application/src/main/java/run/halo/app/plugin/PluginApplicationContext.java:19]: 重点关注一下创建 ApplicationContext 的耗时

重构之后：
- app-store-integration -> 5795 ms
- PluginCommentWidget  -> 1076 ms
- PluginFeed  -> 9404 ms
- PluginS3ObjectStorage  -> 439 ms
- PluginMigrate  -> 486 ms
- PluginSitemap  -> 314 ms

之前：
- app-store-integration -> 39 ms
- PluginCommentWidget -> 1ms
- PluginSearchWidget -> 2ms
- PluginSitemap -> 1ms
- PluginFeed -> 0ms
- PluginMigrate -> 0ms
- PluginS3ObjectStorage -> 1ms

Comment (JohnNiang) [reply application/src/main/java/run/halo/app/plugin/PluginApplicationContext.java:19]: 我在本地测试了一下同一个插件在 Halo 启动之后重新启用，结果如下：

修改之前：

```java
2024-01-25T17:42:38.699+08:00 DEBUG 81162 --- [nReconciler-t-1] r.h.a.p.PluginApplicationInitializer     : initApplicationContext total millis: 59 ms -> StopWatch '': 0.059972791 seconds
----------------------------------------
Seconds       %       Task name
----------------------------------------
0.002254666   04%     createPluginApplicationContext
0.001857      03%     findCandidateComponents
0.000470208   01%     registerBean
0.055390917   92%     refresh plugin application context
```

修改之后：

```java
2024-01-25T17:47:23.417+08:00 DEBUG 81354 --- [nReconciler-t-1] p.DefaultPluginApplicationContextFactory : 
StopWatch 'CreateApplicationContextForapp-store': 67.336207 milliseconds
------------------------------------------------------------------------
Milliseconds  %       Task name
------------------------------------------------------------------------
00.222791     00%     Create
01.160333     02%     LoadPropertySources
00.064291     00%     RegisterBeans
00.987875     01%     LoadComponents
64.900917     96%     Refresh
```

注意时间单位，前者是 `Seconds`，后者是 `Milliseconds`，实际测试差距并不是很大。

Comment (guqing) [reply application/src/main/java/run/halo/app/plugin/PluginApplicationContext.java:19]: 我擦 好奇怪 我怎么再跑了一下就又快了 那之前这么离谱 5s 都干出来了

**Review by guqing — APPROVED** [review]
/approve

未见异常

**Review by ruibaby — APPROVED** [review]
/lgtm

**Pull request merged**